### PR TITLE
Fix #306 - Support importing .zip archives

### LINKED ISSUE #306: Support importing/exporting .zip archives
* Maybe dropbox integration?
 * Just the design
* As zipped files (import/export)
* Drag and drop?
* URL to files?

### PR DESCRIPTION
**Pull request by humphd**
This adds drag-and-drop .zip support, asking the user whether they want the zip file saved or extracted.  I'll do project export to .zip in #347, but I've added 'zip' at the same time as 'unzip' anyway.

### CODE REVIEW

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 1]
```diff
@@ -21,6 +21,70 @@ define(function (require, exports, module) {
...
+                        return callback(err);
+                    } else {
+                        parent = Path.dirname(path);
+                        if(parent === "/") {
+                            proxyFS.mkdir(path, function (err) {
+                                if (err && err.code !== "EEXIST") {
+                                    return callback(err);
+                                }
+
+                                return callback();
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/BracketsFiler.js]: Don't really need return here do we?

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 2]
```diff
@@ -21,6 +21,70 @@ define(function (require, exports, module) {
...
+
+                proxyFS.stat(path, function(err, stat) {
+                    if(stat) {
+                        if(stat.type === "DIRECTORY") {
+                            return callback();
+                        } else if (stat.type === "FILE") {
+                            return callback(filerError("ENOTDIR", path));
+                        }
+                    } else if (err && err.code !== "ENOENT") {
+                        return callback(err);
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/BracketsFiler.js]: or here

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 3]
```diff
@@ -21,6 +21,70 @@ define(function (require, exports, module) {
...
+
+            function _mkdirp(path, callback) {
+                var parent;
+
+                proxyFS.stat(path, function(err, stat) {
+                    if(stat) {
+                        if(stat.type === "DIRECTORY") {
+                            return callback();
+                        } else if (stat.type === "FILE") {
+                            return callback(filerError("ENOTDIR", path));
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/BracketsFiler.js]: or here

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 4]
```diff
@@ -21,6 +21,70 @@ define(function (require, exports, module) {
...
+                return callback();
+            }
+
+            function _mkdirp(path, callback) {
+                var parent;
+
+                proxyFS.stat(path, function(err, stat) {
+                    if(stat) {
+                        if(stat.type === "DIRECTORY") {
+                            return callback();
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/BracketsFiler.js]: or here?

File: src/filesystem/impls/filer/BracketsFiler.js [hunk 5]
```diff
@@ -21,6 +21,70 @@ define(function (require, exports, module) {
...
+                                if (err) {
+                                    return callback(err);
+                                }
+
+                                proxyFS.mkdir(path, function (err) {
+                                    if (err && err.code !== "EEXIST") {
+                                        return callback(err);
+                                    }
+
+                                    return callback();
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/BracketsFiler.js]: or here?

File: src/utils/DragAndDrop.js [hunk 6]
```diff
@@ -181,6 +182,68 @@ define(function (require, exports, module) {
...
             var errorList = [];
 
+            function handleRegularFile(deferred, file, filename, buffer, encoding) {
+                file.write(buffer, {encoding: encoding}, function(err) {
+                    if (err) {
+                        deferred.reject(err);
+                        return;
+                    }
+
+                    if(encoding === "utf8") {
```
Comment (gideonthomas) [inline src/utils/DragAndDrop.js]: Why are we doing this check?

**Comment by gideonthomas** [comment]
this is going to be a great feature!

File: src/utils/DragAndDrop.js [hunk 7]
```diff
@@ -181,6 +182,68 @@ define(function (require, exports, module) {
...
+                                }
+
+                                Dialogs.showModalDialog(
+                                    DefaultDialogs.DIALOG_ID_INFO,
+                                    "Unzip Completed Successfully",
+                                    "Successfully unzipped <b>" + basename + "</b>."
+                                )
+                                    .done(function() {
+                                        deferred.resolve();
+                                    });
```
Comment (sedge) [inline src/utils/DragAndDrop.js]: Probably a good idea to add a `.catch()` here for modal errors.

File: src/utils/DragAndDrop.js [hunk 8]
```diff
@@ -181,6 +182,68 @@ define(function (require, exports, module) {
...
+                            });
+                        } else {
+                            handleRegularFile(deferred, file, filename, buffer, encoding);
+                        }
+                    });
```
Comment (sedge) [inline src/utils/DragAndDrop.js]: and here

File: src/utils/DragAndDrop.js (hunk 6 above)
Comment (humphd) [inline src/utils/DragAndDrop.js]: So that I don't try to open a non-UTF8 file in the editor (e.g., the .zip file).

Comment (gideonthomas) [inline src/utils/DragAndDrop.js]: So how would this work for images?

**Comment by gideonthomas** [comment]
looks good! r=me

**Comment by sedge** [comment]
r+

File: src/utils/DragAndDrop.js (hunk 6 above)
Comment (humphd) [inline src/utils/DragAndDrop.js]: We just won't automatically open them, but they'll be in the tree.  If you want me to do a check for any image type, and override, I could.  How much does it matter?

**Pull request merged**